test/provider_devdax_memory.cpp

```diff
@@ -0,0 +1,243 @@
+// Copyright (C) 2024 Intel Corporation
+// Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "base.hpp"
+
+#include "cpp_helpers.hpp"
+
+#include <umf/memory_provider.h>
+#include <umf/providers/provider_devdax_memory.h>
+
+using umf_test::test;
+
+#define INVALID_PTR ((void *)0x01)
+
+#define ASSERT_IS_ALIGNED(ptr, alignment)                                      \
+    ASSERT_EQ(((uintptr_t)ptr % alignment), 0)
+
+typedef enum purge_t {
+    PURGE_NONE = 0,
+    PURGE_LAZY = 1,
+    PURGE_FORCE = 2,
+} purge_t;
+
+static const char *Native_error_str[] = {
+    "success",                          // UMF_DEVDAX_RESULT_SUCCESS
+    "memory allocation failed",         // UMF_DEVDAX_RESULT_ERROR_ALLOC_FAILED
+    "allocated address is not aligned", // UMF_DEVDAX_RESULT_ERROR_ADDRESS_NOT_ALIGNED
+    "memory deallocation failed",       // UMF_DEVDAX_RESULT_ERROR_FREE_FAILED
+    "lazy purging failed",  // UMF_DEVDAX_RESULT_ERROR_PURGE_LAZY_FAILED
+    "force purging failed", // UMF_DEVDAX_RESULT_ERROR_PURGE_FORCE_FAILED
+};
+
+// test helpers
+
+static int compare_native_error_str(const char *message, int error) {
+    const char *error_str = Native_error_str[error - UMF_DEVDAX_RESULT_SUCCESS];
+    size_t len = strlen(error_str);
+    return strncmp(message, error_str, len);
+}
+
+using providerCreateExtParams = std::tuple<umf_memory_provider_ops_t *, void *>;
+
+umf::provider_unique_handle_t
+providerCreateExt(providerCreateExtParams params) {
```

Comment:
Please return handle thru value and return void. Expect will not fail test so it will go on, and create multiple other failures in this test - which are misleading during debug.

also calls to this function surround ASSERT_NO_FATAL_FAILURE to propagate failure to test function

Reply:
I believe it can be changed later in a separate PR.

Reply:
I meant I believe it can be changed later in a separate PR, because it is done exactly like in the OS memory provider:
https://github.com/oneapi-src/unified-memory-framework/blob/main/test/provider_os_memory.cpp#L44-L47
So IMHO it can be changed in a separate PR in both places in the same PR.

Reply:
This PR concentrates on adding the devdax provider, not on fixing tests of other providers. Tests of other providers should be fixed in a separate PR.

Reply:
@lplewa Is it enough https://github.com/oneapi-src/unified-memory-framework/pull/719/commits/618f755b0676bc3310be14b28394c8e47d88073d ? (PR https://github.com/oneapi-src/unified-memory-framework/pull/719)

Reply:
Added to this PR.
@lplewa Done